ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java

```diff
@@ -418,16 +459,19 @@ void addEntry(IndexEntry entry, Offset fileEndOffset) throws StoreException {
               + "originalMessageOffset {} fileEndOffset {}", indexFile.getAbsolutePath(), entry.getKey(),
           entry.getValue().getOffset(), entry.getValue().getSize(), entry.getValue().getExpiresAtMs(),
           entry.getValue().getOriginalMessageOffset(), fileEndOffset);
-      if (index.put(entry.getKey(), entry.getValue()) == null) {
-        numberOfItems.incrementAndGet();
-        sizeWritten.addAndGet(entry.getKey().sizeInBytes() + entry.getValue().getBytes().capacity());
+      boolean isPresent = index.containsKey(entry.getKey());
+      index.computeIfAbsent(entry.getKey(), key -> new ConcurrentSkipListSet<>(INDEX_VALUE_COMPARATOR))
+          .add(entry.getValue());
+      if (!isPresent) {
         bloomFilter.add(ByteBuffer.wrap(entry.getKey().toBytes()));
-        if (resetKey == null) {
-          resetKey = new Pair<>(entry.getKey(),
-              entry.getValue().isFlagSet(IndexValue.Flags.Delete_Index) ? PersistentIndex.IndexEntryType.DELETE
-                  : PersistentIndex.IndexEntryType.PUT);
-        }
       }
+      if (resetKey == null) {
+        resetKey = new Pair<>(entry.getKey(),
+            entry.getValue().isFlagSet(IndexValue.Flags.Delete_Index) ? PersistentIndex.IndexEntryType.DELETE
```

Comment:
Do you mean `resetKey`? Nothing today - it was meant to serve as a hint about where to resume replication from if compaction occurred (never implemented). Don't need a `TTL` type since 1) does not affect correctness to set it to PUT (will just resume further down than it should) 2) Without revisiting why and how `resetKey` will be used, doesn't make sense to make and accommodate changes here.